text/1003-cli-naming.rst

```diff
@@ -0,0 +1,258 @@
+adopt the ``<group> <command>`` for all other commands.  Combining two
+drastically different CLI designs in one tool is not an option.
+
+
+Overview
+========
+
+The RFC proposes to adopt the ``<group> <command>`` naming scheme for
+the ``edgedb`` CLI.  Here's an outline of current and future ``edgedb``
+subcommands:
+
+* ``server`` (see also RFC 1001.)
+  - ``init``
+  - ``install``
+  - ``restart``
+  - ``start``
+  - ``status``
+  - ``stop``
+
+* ``dump``
+  - ``db`` -- backup a database.
+  - ``all`` -- backup all databases, as well as roles, configs, etc into
+    a directory.
+  - ``restore-db``
+  - ``restore-all``
+  - ``config`` -- backup system configuration.
+
+* ``migration``
+  - ``status``
+  - ``create``
+  - ``apply``
+
+* ``role``
+  - ``create [--superuser]``
+  - ``alter``
+  - ``drop``
+  - ``list``
+
+* ``db``
+  - ``create``
+  - ``rename``
+  - ``drop``
+  - ``list``
+
+* ``config [--system]`` (used to be ``edgedb configure``)
+  - ``set``
+  - ``reset``
+  - ``add`` (used to be ``insert``)
+  - ``show``
+
+* ``run [--stdin | -c]`` -- run an EdgeQL script or command from stdin
+  or passed via the command line with ``-c``.
+
+
+Design Considerations
+=====================
+
+List Commands
+-------------
+
+Currently, EdgeDB REPL has shortcuts to list aliases, roles, etc. for the
+current database, with ``\la`` or ``\list-aliases`` kind of syntax.  The
+list commands are also exposed to the CLI.
+
+While these commands are quite handy to have inside REPL, their usefulness
+as standalone CLI tools is questionable, especially when an arbitrary EdgeQL
+introspection query (or even shortcuts like ``\la``) can be easily piped into
+the ``edgedb`` command.
+
+This RFC proposes to limit the number of actual CLI commands to the practical
+minimum.
+
+
+CLI via REPL
+------------
+
+REPL-specific commands should be exposed via the ``\`` prefix, e.g.::
+
+  \d [-v] NAME             describe schema object
+  \l, \list-databases      list databases
+  \lT [-sI] [PATTERN]      list scalar types
+                           (alias: \list-scalar-types)
+  <...>
+
+It is convenient to expose the EdgeDB CLI commands directly in REPL.  This
+can greatly simplify administrative tasks when the DB administrator has full
+DB admin rights but yet can't access the server shell.  For that purpose, all
+CLI commands are exposed with the ``!`` prefix, similar to how IPython
+exposes shell commands::
+
+  >>> \list-databases
+  tutorial
+
+  >>> SELECT 1;
+  {1}
+
+  >>> !dump db tutorial
+  done
+
+This way the internal REPL help system is not overloaded with rarely
+needed help on CLI commands and would only show the list of convenient ``\``
+commands with a hint that ``!help`` or ``!h`` can be used to list
+all CLI options.
+
+Name-spacing ``\help`` and ``!help`` is good for the usability, because
+the latter set of commands is not going to be used as frequently as
+the former.
+
+
+Connection Options
+------------------
```

Comment:
Second problem is these arguments:
```             
   --password                   Ask for password on the terminal (TTY)                                                                                         
   --password-from-stdin        Read the password from stdin rather than TTY (useful for scripts)
```
Because they are technically also connection arguments. And same names are reused in ``create role`` / ``alter role``.

Reply:
Third problem is that it's unsupported in clap: https://github.com/clap-rs/clap/issues/1385
(even if when it becomes supported I'm not sure about the granularity of that setting, as it may apply to all arguments rather than selected ones). The implementation of the hoisting connection arguments up would require another piece of code for each command.

Surely, we can try to contribute the feature, but given the scope of clap (i.e. all kind of completion generation and so on) I'm not sure how hard it would be to make a patch and have it upstreamed.

Of course, these technical issues are fixable eventually, just noting that it isn't a very simple task.

Reply:
Another issue is if we go with `dump all` then `dump all --help` will list `-d database` which is quite weird.